mesonbuild/interpreter/interpreterobjects.py

```diff
@@ -83,7 +83,7 @@ def extract_search_dirs(kwargs: 'kwargs.ExtractSearchDirs') -> T.List[str]:
 class FeatureOptionHolder(ObjectHolder[coredata.UserFeatureOption]):
     def __init__(self, option: coredata.UserFeatureOption, interpreter: 'Interpreter'):
         super().__init__(option, interpreter)
-        if option and option.is_auto():
+        if option and option.is_auto() and not option.is_user_input():
```

Comment:
I think you can instead do `option.name in interpreter.user_defined_options.cmd_line_options`. However, I'm wondering what we should do for options that have been set to `auto` in the cross/native file, those should be considered set by user too afaik. tbf, I'm not entirely happy with `interpreter.user_defined_options` it should be refactored as a boolean on the UserOption object as you tried to do indeed.